### ISSUE TITLE
v1.1.0 crash with UIView margins

### ISSUE THREAD

**Issue by CristianoCastro** [SnapKit/Masonry#461]
### New Issue Checklist

🚫 If this template is not filled out your issue **will** be closed with no comment. 🚫

* [ ] I have looked at the [Documentation](https://github.com/SnapKit/Masonry/blob/master/README.md)
* [ ] I have filled out this issue template.

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform                | ios
 Platform Version        | 9.X & 10.X
 Masonry Version         | 1.1.0
 Integration Method      | cocoapods
 

### Issue Description

Crash: `-[UIView mas_bottomMargin]: unrecognized selector sent to instance 0x7fe34460a6a0`
Possibly other margin methods too.

Worked with Masonry v1.0.2

**Comment by beloso** [comment]
It happens with me too

**Comment by robertjpayne** [comment]
This doesn't make much sense as the code here has hardly changed at all… Does it work with Xcode 9 and Masonry 1.0.2?

**Comment by guofengld** [comment]
Xcode 9.0 + Masonry 1.1.0 crashes,  Clean -> Build, then goes OK, confused.

**Comment by robertjpayne** [comment]
@guofengld so it doesn't crash once you do a full clean + build? Sounds like a weird Xcode issue

**Comment by jonz-tech** [comment]
xcode 8.3.3 build the lasted version will cash when use XXXMargin . so i don't think so it's a weird xcode issuse.

**Comment by robertjpayne** [comment]
@jonz-tech have you tried a full clean? Including deleting your DerrivedData folder? It seems super suspicious because nothing about these API's changed.

**Comment by guofengld** [comment]
@robertjpayne "ios deployment_target" was set to "6.0", so __IPHONE_OS_VERSION_MIN_REQUIRED equals to 60000, mas_leftMargin、mas_rightMargin were not implemented, but they were declared 😂

**Comment by jonz-tech** [comment]
i had try. here is my issue. 

https://github.com/SnapKit/Masonry/issues/460

i had checked your project .it seems nothing change, and still cause some problem.
does the implement was missing in podspest?

**Comment by CristianoCastro** [comment]
Updated issue to add some info about Xcodes and devices tested

**Comment by robertjpayne** [comment]
@CristianoCastro super helpful thank you, I have no idea why it's letting the header go through but not the API, gah!

**Comment by robertjpayne** [comment]
@CristianoCastro just one last question -- does it crash on device? Or just simulators?

**Comment by robertjpayne** [comment]
Hrm I'm not really sure what to say here, if I clone this project and run the Masonry iOS Examples and tap into the `Margins` example it doesn't crash.

This is with shorthand mode on.

Whats your min deployment target set to?

**Comment by beloso** [comment]
It crashes on simulators as well. We are testing our UI on the iPhone X simulator and it crashes there.

**Comment by robertjpayne** [comment]
@beloso can you try just downloading the .tar.gz of the v1.1.0 release and running the examples and tapping into the margins example?

I can't replicate the crash on my end…

**Comment by CristianoCastro** [comment]
@robertjpayne updated issue to include results of running the example `Margins` and a test project

**Comment by robertjpayne** [comment]
@CristianoCastro If I had to bet I'd say it's Cocoapods and Xcode 9, are you using static libraries or frameworks with Cocoapods?

**Comment by CristianoCastro** [comment]
@robertjpayne it also happens with XCode 8.3.3. I'm not using frameworks.

Another difference I've noticed is that the Example `Margins` uses `[UIView bottomMargin]` and not `[UIView mas_bottomMargin]` as I am:
`make.bottom.equalTo(lastView.bottomMargin);`

**Comment by CristianoCastro** [comment]
Updated issue to include sample project that crashes

**Comment by robertjpayne** [comment]
@CristianoCastro hm yea so if you don't `use_frameworks!` with Cocoapods it doesn't load all the symbols. I'll have to see why these symbols are not being loaded…

**Comment by robertjpayne** [comment]
So it appears if you use:

```
#if TARGET_OS_IPHONE || TARGET_OS_TV
```

The symbols are exported in a framework and a static lib properly. However if you use:

```
#if (__IPHONE_OS_VERSION_MIN_REQUIRED >= 80000) || (__TV_OS_VERSION_MIN_REQUIRED >= 9000)
```

The symbols are not properly exported in a static lib.

The work around right now is to `use_frameworks!`

**Comment by robertjpayne** [comment]
I'm not 100% sure how to work around this - the reason the API's were changed is due to new warnings Xcode gives around the use of potentially unavailable attributes

**Comment by CristianoCastro** [comment]
unfortunately for me I can't `use_frameworks!` because some pods don't play nice with it.
I will see if I can find a better way

**Comment by CristianoCastro** [comment]
updated issue with a workaround other than `use_frameworks!`

**Comment by robertjpayne** [comment]
@CristianoCastro Interesting, I wonder why Cocoapods is setting deployment targets to iOS 6…

**Comment by robertjpayne** [comment]
Probably because we have it in the pod spec, but Cocoapods should override this to whatever is higher :(

**Comment by teerapap** [comment]
These links below may explain CocoaPods's behaviour.  It seems like it does something different with `use_framework!`.
CocoaPods/CocoaPods#3454
CocoaPods/CocoaPods#4320

**Comment by robertjpayne** [comment]
@teerapap yea seems like a Cocoapods bug more than anything, I don't get why it isn't setting the min deployment target to the main app target…

**Comment by robertjpayne** [comment]
Can people please try to use master and let me know if this resolves the issue(s)? I reverted some availability macros (stuff may crash now rather than not work) and used Xcode 9's new `NS_AVAILABLE` macros where possible to allow for warnings.

**Comment by CristianoCastro** [comment]
@robertjpayne the test project is working now when using the master branch, tested in XCode 9 on iOS 10 & 11. Though I have warnings about some enumerations in masonry being partial (I don't recall if they are new or not)

<img width="1261" alt="screen shot 2017-09-28 at 14 33 38" src="https://user-images.githubusercontent.com/6043612/30969376-1b04614c-a45a-11e7-8fdd-e9f1d56a415f.png">

**Comment by hhyymm** [comment]
cocoapods 
Masonry Version | 1.1.0  
 iOS Simulator 8.4 && 11.0
Xcode 8.3.3 + Xcode9.0 

Podfile use setting blow

post_install do |installer|
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings['IPHONEOS_DEPLOYMENT_TARGET'] = '8.0'
    end
  end
end

It works well

**Comment by robertjpayne** [comment]
@CristianoCastro did you ever find a way around the warnings?

**Comment by CristianoCastro** [comment]
@robertjpayne I didn't research the subject further due to a deadline and the cocoapods workaround of setting each target's IPHONEOS_DEPLOYMENT_TARGET doing the trick

**Comment by xingheng** [comment]
I don't think this is an issue related cocoapods, the cocoapods uses the main app project's deployment target when running `pod install` at the first time only, AFAIK. However, this library's deployment target is iOS 6.0 in the podspec file, then the precompiler should support all the above iOS systems. The margins (`leftMargins` and `bottomMargin`) is supported on iOS 8 only, those symbols shouldn't be occurred in the main project code which has an iOS 6/7 deployment target. It should be a kind of compiling error there.

The new introduced macros in 1.1.0 get the hint.

`#if TARGET_OS_IPHONE || TARGET_OS_TV`

@robertjpayne

**Comment by Rogerlin2013** [comment]
@xingheng Can't agree you more

**Comment by godera** [comment]
@robertjpayne when adding files to project, i take no error. But podding it cause 'unrecognized selector sent to instance xxx'.

**Comment by Lucashuang0802** [comment]
@xingheng so do we have a solution here?

**Comment by Suncoldx333** [comment]
<img width="377" alt="wx20180906-153509 2x" src="https://user-images.githubusercontent.com/15140575/45142192-79479c80-b1ea-11e8-9a95-1d58d2471044.png">

by command "pod install --verbose",  I find cocoapods install masonry with 6.0 which decided in Masonry.podspec 

<img width="136" alt="wx20180906-154302 2x" src="https://user-images.githubusercontent.com/15140575/45142630-94ff7280-b1eb-11e8-86ed-9053c6d5f0ac.png">

so click Pods->Masonry,  search "deployment target" in "Build Settings", check if it is 6.0, and choose version you want

<img width="1116" alt="wx20180906-155300 2x" src="https://user-images.githubusercontent.com/15140575/45143322-8ade7380-b1ed-11e8-8ff1-051249501763.png">

**Comment by ghost** [comment]
This one fixes the Crash, Thanks! @Suncoldx333 

The iOS Deployment Target is changed to "iOS 6.0", Change back to 8.0 or higher works for me

**Comment by xingheng** [comment]
@Suncoldx333 I don't think that's a good fix, the `Pods/Pods.xcodeproj` is under the `Pods` directory which is added to `.gitignore` in SCM in general, because `cocoapods` manages the project file via automation completely.

**Comment by Suncoldx333** [comment]
> @Suncoldx333 I don't think that's a good fix, the `Pods/Pods.xcodeproj` is under the `Pods` directory which is added to `.gitignore` in SCM in general, because `cocoapods` manages the project file via automation completely.

yes you are right. what I did will be nothing after pod install.  and I think this issue has been solved when @CristianoCastro share the Edit 4, just add those code into podfile.

**Comment by ghost** [comment]
I notice there is a **Waring** for my project after pods install, Xcode 10 will automatically change 
iOS deployment target to the right one.
![image](https://user-images.githubusercontent.com/34498461/50476139-f1d4a200-0a01-11e9-9a90-17f423b77f88.png)

![image](https://user-images.githubusercontent.com/34498461/50476131-eda88480-0a01-11e9-835b-83ced07c73ce.png)

**Comment by RbBtSn0w** [comment]
So, how about now?

**Comment by desperadocarl** [comment]
So, how about now? confused！
